File: apis/edge-api/admin/src/delivery/pricing.ts
// Service layer for delivery PRICING RULES (032) — validation and orchestration. No HTTP, no SQL
// (Principle VI). Tests mock ./pricing-repository at the module boundary.
//
// ⚠ THE REFUSALS ARE THE POINT OF THIS FILE. A pricing rule is operator-editable configuration that
// silently decides what every shopper pays, and the failure modes are all quiet ones — a cap below
// the floor makes every fee the cap; an empty band set prices everything at the base; an unrounded
// cap produces an odd number only on the most expensive orders. None of those throws anything at
// runtime. They just charge the wrong amount, forever, until somebody notices.
//
// Each refusal therefore has its own stable code, because "invalid" tells an operator nothing about
// which of the rules they broke. Same reasoning as 027's eight distinguishable promo refusals.
import * as repo from "./pricing-repository";
import {
  type DeliveryMethod,
  DELIVERY_METHODS,
  DeliveryError,
  type DeliveryRefusalCode,
  type FieldIssue,
  type PriceBand,
  type PricingRule,
  type PricingRuleInput,
} from "./types";

const AMOUNT_RE = /^\d+(\.\d{1,2})?$/;
const BOUND_RE = /^\d+(\.\d{1,2})?$/;

export function listRules(): Promise<PricingRule[]> {
  return repo.listRules();
}

/** Replace one method's rule and its whole band set. */
export async function replaceRule(
  methodRaw: string,
  body: Record<string, unknown>,
  actorSub: string,
): Promise<PricingRule> {
  if (!(DELIVERY_METHODS as readonly string[]).includes(methodRaw)) {
    throw new DeliveryError("not_found", `unknown delivery method ${methodRaw}`);
  }
  const method = methodRaw as DeliveryMethod;

  const fields: FieldIssue[] = [];
  const baseAmount = amount(body.baseAmount, "baseAmount", fields);
  const roundingStep = amount(body.roundingStep, "roundingStep", fields);
  const maxAmount = amount(body.maxAmount, "maxAmount", fields);
  const status = body.status === "disabled" ? "disabled" : "active";

  const distanceBands = bands(body.distanceBands, "distanceBands", fields);
  const weightBands = bands(body.weightBands, "weightBands", fields);

  if (fields.length > 0) {
    throw new DeliveryError("validation", "invalid pricing rule", fields);
  }

  // ── The five semantic refusals ──────────────────────────────────────────────────────────────
  //
  // Everything above is "is this a number?". Everything below is "would this configuration do
  // something the operator did not intend?", which is why these are 422 and carry a code.

  // ⚠ An empty band set is not "no adjustment" — it silently prices EVERY distance and EVERY weight
  // at the base, which is exactly the FR-011 defect (a gap meaning no fee) applied to the whole axis.
  if (distanceBands.length === 0 || weightBands.length === 0) {
    throw refuse(
      "bands_required",
      "both a distance band set and a weight band set are required — without them every order prices at the base amount, whatever it weighs and however far it goes",
    );
  }

  // Two answers for one value. Which one applies would come down to row order.
  for (const [label, set] of [["distance", distanceBands], ["weight", weightBands]] as const) {
    const seen = new Set<string>();
    for (const b of set) {
      const key = String(Number(b.upperBound));
      if (seen.has(key)) {
        throw refuse("duplicate_band", `two ${label} bands share the upper bound ${b.upperBound}`);
      }
      seen.add(key);
    }
  }

  const step = Number(roundingStep);
  if (!(step > 0)) {
    throw refuse("invalid_rounding", "the rounding step must be greater than zero");
  }

  // ⚠ THE FLOOR IS BASE + THE SMALLEST BAND FROM EACH AXIS, NOT THE LARGEST.
  //
  // Comparing against the largest bands would refuse every cap that could ever bind — which is
  // precisely what a cap is FOR. The real defect a cap-below-floor produces is a silently FLAT price
  // table: if the cheapest possible order already exceeds the ceiling, every order costs the ceiling,
  // distance and weight stop mattering entirely, and nothing reports it.
  const floor = Number(baseAmount) + smallestAdd(distanceBands) + smallestAdd(weightBands);
  const cap = Number(maxAmount);
  if (cap < floor) {
    throw refuse(
      "cap_below_floor",
      `the maximum ${maxAmount} is below the cheapest possible fee (${floor.toFixed(2)}) — every delivery would cost the maximum, and distance and weight would stop affecting the price at all`,
    );
  }

  // ⚠ min(cap, roundUp(...)) returns the cap verbatim. If the cap is not itself a multiple of the
  // step, the fee is unrounded at exactly the moment the cap binds — on the most expensive orders,
  // where an odd number is least likely to be spotted. SC-003 says 100% of fees are rounded.
  if (!isMultiple(cap, step)) {
    throw refuse(
      "cap_not_rounded",
      `the maximum ${maxAmount} is not a multiple of the rounding step ${roundingStep} — capped fees would not be rounded`,
    );
  }

  const input: PricingRuleInput = {
    baseAmount: baseAmount!,
    roundingStep: roundingStep!,
    maxAmount: maxAmount!,
    status,
    distanceBands,
    weightBands,
  };
  return repo.replaceRule(method, input, actorSub);
}

function refuse(code: DeliveryRefusalCode, message: string): DeliveryError {
  return new DeliveryError("unprocessable", message, undefined, code);
}

function smallestAdd(set: PriceBand[]): number {
  return set.reduce((min, b) => Math.min(min, Number(b.addAmount)), Number.POSITIVE_INFINITY);
}

/**
 * Multiple-of check in integer cents.
 *
 * ⚠ `45.00 % 0.50` in binary floating point is not reliably 0 — this is the classic case where a
 * modulus on floats reports a false failure and an operator cannot save a perfectly valid ceiling.
 * Both sides are money with at most two decimals, so integer cents is exact.
 */
function isMultiple(value: number, step: number): boolean {
  const v = Math.round(value * 100);
  const s = Math.round(step * 100);
  return s > 0 && v % s === 0;
}

function amount(raw: unknown, field: string, fields: FieldIssue[]): string | null {
  const s = typeof raw === "number" ? raw.toFixed(2) : typeof raw === "string" ? raw.trim() : "";
  if (!AMOUNT_RE.test(s)) {
    fields.push({ field, message: "must be a non-negative amount with up to 2 decimals" });
    return null;
  }
  return s;
}

function bands(raw: unknown, field: string, fields: FieldIssue[]): PriceBand[] {
  if (raw === undefined || raw === null) return [];
  if (!Array.isArray(raw)) {
    fields.push({ field, message: "must be an array of bands" });
    return [];
  }
  const out: PriceBand[] = [];
  for (const [i, item] of raw.entries()) {
    const row = item as Record<string, unknown>;
    const ub = typeof row?.upperBound === "number" ? String(row.upperBound) : String(row?.upperBound ?? "").trim();
    const add = amount(row?.addAmount, `${field}[${i}].addAmount`, fields);
    if (!BOUND_RE.test(ub) || Number(ub) <= 0) {
      fields.push({ field: `${field}[${i}].upperBound`, message: "must be a positive number" });
      continue;
    }
    if (add === null) continue;
    out.push({ upperBound: ub, addAmount: add });
  }
  // Sorted here so the stored set is always ascending regardless of how a console sent it — the
  // pricing core sorts too, but a table an operator reads should not be in submission order.
  out.sort((a, b) => Number(a.upperBound) - Number(b.upperBound));
  return out;
}

File: apis/core-api/internal/platform/delivery/pricing.go
package delivery

import "sort"

// Band is one step of a pricing rule: everything up to UpperBound adds AddCents.
//
// ⚠ UPPER BOUND ONLY. Storing both bounds would make a GAP between two bands REPRESENTABLE, and a gap
// is the defect FR-011 exists to prevent — a distance or weight matching no band must never mean "no
// fee". With only an upper bound there is nothing to fall between.
type Band struct {
	UpperBound float64 // kilometres, or kilograms
	AddCents   int64
}

// PricingRule is how one delivery method is priced. Loaded from public.delivery_pricing_rule.
type PricingRule struct {
	Method          Method
	BaseCents       int64
	RoundingStepCts int64
	MaxCents        int64
	DistanceBands   []Band
	WeightBands     []Band
	Active          bool
}

// Price computes the fee for one package, in integer cents.
//
//	fee = min(cap, roundUp(base + distanceBand + weightBand, step))
//
// It returns ok=false only when the method is not offered at all (a disabled rule). ⚠ It NEVER returns
// ok=false because an input was out of range: every distance and every weight has a defined answer,
// because a gap in operator-editable configuration must not refuse a sale or give delivery away.
//
// The rules, each of which is a test in pricing_test.go:
//
//	distance beyond the last band  -> the last band     (not free, not undeliverable)
//	weight beyond the last band    -> the last band
//	distance UNKNOWN               -> the FURTHEST band  (FR-038 — never the nearest)
//	no bands configured at all     -> the base alone
//
// ⚠ The unknown-distance rule is the one that matters. `distanceKnown` is false when either the shop's
// or the shopper's postcode has no centroid. Treating that as 0 km would make the most remote postcode
// in the country — precisely the one whose location we are least likely to have — the cheapest to
// deliver to, with nothing anywhere reporting a fault. Pricing it as the furthest band is the safe
// direction to be wrong in.
func Price(rule PricingRule, distanceKm float64, distanceKnown bool, weightKg float64) (int64, bool) {
	if !rule.Active {
		return 0, false
	}

	cents := rule.BaseCents + bandAdd(rule.DistanceBands, distanceKm, distanceKnown) + bandAdd(rule.WeightBands, weightKg, true)
	cents = roundUpTo(cents, rule.RoundingStepCts)

	// ⚠ The cap rounds DOWN to the step, never up past itself: a ceiling the operator set must not be
	// exceeded by the rounding that was supposed to tidy the number. The schema requires the cap to be
	// a multiple of the step anyway (delivery_pricing_rule_cap_rounded_ck), so this only bites if that
	// constraint is ever relaxed — but a capped fee that is not a round number breaks SC-003 on exactly
	// the most expensive orders, where nobody is looking.
	if rule.MaxCents > 0 && cents > rule.MaxCents {
		cents = roundDownTo(rule.MaxCents, rule.RoundingStepCts)
	}
	return cents, true
}

// bandAdd picks the smallest band whose UpperBound is >= value.
//
// ⚠ `known=false` deliberately skips the search and takes the LAST (largest) band. See Price.
func bandAdd(bands []Band, value float64, known bool) int64 {
	if len(bands) == 0 {
		return 0
	}
	sorted := make([]Band, len(bands))
	copy(sorted, bands)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].UpperBound < sorted[j].UpperBound })

	if !known {
		return sorted[len(sorted)-1].AddCents
	}
	for _, b := range sorted {
		// ⚠ >= , so a value EXACTLY on a bound takes that band. `>` would push every boundary value
		// into the next band up, shifting prices at every band edge in a way an aggregate test would
		// not catch.
		if value <= b.UpperBound {
			return b.AddCents
		}
	}
	// Beyond every band -> the largest. NOT zero: that would be free delivery to the far side of the
	// country, produced by the absence of a row.
	return sorted[len(sorted)-1].AddCents
}

// roundUpTo rounds to the next multiple of step.
//
// ⚠ UP, NEVER NEAREST (FR-005). Nearest-rounding means the platform silently absorbs the difference on
// roughly half of all orders — a revenue decision wearing a formatting choice's clothes.
func roundUpTo(cents, step int64) int64 {
	if step <= 0 {
		return cents
	}
	if r := cents % step; r != 0 {
		cents += step - r
	}
	return cents
}

func roundDownTo(cents, step int64) int64 {
	if step <= 0 {
		return cents
	}
	return cents - cents%step
}
